### PR TITLE
Containerise app

### PR DESCRIPTION
**Pull request by thirionlogan**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM node:latest
+
+WORKDIR /usr/src/app
+
+ENV INLINE_RUNTIME_CHUNK false
```
Comment (thirionlogan) [inline Dockerfile:5]: prevents webpack from injecting unsafe inline scripts into the HTML

File: public/index.html [hunk 2]
```diff
@@ -3,7 +3,7 @@
   <head>
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1" />
-    <meta name="theme-color" content="#000000" />
+    <meta property="csp-nonce" content="{{ styleNonce }}" />
```
Comment (thirionlogan) [inline public/index.html:6]: inject a nonce for certifying style elements generated by a material-ui dependency (JSS)

File: server/knexfile.js [hunk 3]
```diff
@@ -2,7 +2,7 @@ require('dotenv').config({ path: '../.env' });
 module.exports = {
   development: {
     client: 'pg',
-    connection: {
+    connection: process.env.DATABASE_URL || {
```
Comment (thirionlogan) [inline server/knexfile.js:5]: inject DB URL string if it exists

File: src/client/mockClient.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    return [
+      `mock${
+        clientFunctionName.charAt(0).toUpperCase() + clientFunctionName.slice(1)
+      }`,
+      jest.fn(clientFunction),
+    ];
+  })
+);
+
+module.exports = mockClient;
```
Comment (thirionlogan) [inline src/client/mockClient.js:14]: Automatically wrap all of the client functions in a jest wrapper for testing

File: src/client/client.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+const instance = axios.create({
+  withCredentials: true,
+  baseURL: API_SERVER,
+});
+
+module.exports = {
+  ping: () => {
+    return instance.get('/');
+  },
+};
```
Comment (thirionlogan) [inline src/client/client.js:13]: this client module will hold all of the functions that call the back end

File: server/app/app.js [hunk 6]
```diff
@@ -6,10 +6,41 @@ const session = require('express-session');
...
-app.use(helmet());
+nunjucks.configure('build', {
+  autoescape: true,
+  express: app,
+});
+
+app.use((req, res, next) => {
+  res.locals.styleNonce = Buffer.from(uuidv4()).toString('base64');
+  next();
+});
```
Comment (thirionlogan) [inline server/app/app.js:23]: Generate a hash for a nonce to certify injected elements

File: server/app/app.js [hunk 7]
```diff
@@ -6,10 +6,41 @@ const session = require('express-session');
...
+    contentSecurityPolicy: {
+      directives: {
+        ...helmet.contentSecurityPolicy.getDefaultDirectives(),
+        'default-src': ["'self'"],
+        'script-src': ["'self'"],
+        'style-src': [
+          "'self'",
+          'fonts.googleapis.com',
+          'fonts.gstatic.com',
+          (req, res) => `'nonce-${res.locals.styleNonce}'`,
```
Comment (thirionlogan) [inline server/app/app.js:36]: insert nonce into allowed style sources

File: server/app/app.js [hunk 8]
```diff
@@ -25,9 +56,33 @@ app.use(
...
+    exposedHeaders: ['set-cookie'],
+  })
+);
+
+app.use('/static', express.static(path.resolve('./build/static')));
+app.use(
+  '/asset-manifest.json',
+  express.static(path.resolve('./build/asset-manifest.json'))
+);
+app.use('/robots.txt', express.static(path.resolve('./build/robots.txt')));
```
Comment (thirionlogan) [inline server/app/app.js:77]: Provide static files to the client

File: server/app/app.js [hunk 9]
```diff
@@ -25,9 +56,33 @@ app.use(
...
+
+app.get('/', (req, res) => {
+  res.render('index.html', {
+    styleNonce: res.locals.styleNonce,
+  });
+});
```
Comment (thirionlogan) [inline server/app/app.js:83]: inject nonce into html and serve the file

**Review by knieszj — APPROVED** [review]
5 of 5 stars.

**Review by adder206 — APPROVED** [review]

**Pull request merged**